### PR TITLE
Validate containers in parallel.

### LINKED ISSUE #272: Parallelize validation checks over containers
**Description**

Policy validation [currently checks each container one at a time](https://github.com/sigstore/policy-controller/blob/4963d61971e06c2543019cc3737ee9dfdc4c5cea/pkg/webhook/validator.go#L164). These checks can involve a lot of blocking IO requesting signatures from repositories and fetching verification keys and such so with a pod with several containers, the latency on validation can grow to be quite high. Would be nice to parallelize these checks over containers similar to the [pattern already used to parallelize the checks over policies](https://github.com/sigstore/policy-controller/blob/4963d61971e06c2543019cc3737ee9dfdc4c5cea/pkg/webhook/validator.go#L301-L312)

### PR DESCRIPTION
**Pull request by vaikas**
Signed-off-by: Ville Aikas <vaikas@chainguard.dev>

#### Summary
Inspired by this comment: https://github.com/sigstore/policy-controller/pull/270#discussion_r982764535

Started refactoring the largish function into smaller pieces and added parallel container validation.
Fixes #272 

Also fix a bug where if requests get cancelled, we'd return before waiting for all goroutines to finish.

#### Release Note

* Validate containers in parallel which will speed up webhook execution for resources with multiple containers.
* Fix an issue where if a context was cancelled before all validations finished, we did not wait for go routines to finish.

### CODE REVIEW

**Review by cpanato — DISMISSED** [review]

**Comment by hectorj2f** [comment]
I am not sure, if we want to wait for https://github.com/sigstore/policy-controller/pull/270 before, let's see if the checks are passing.

**Comment by hectorj2f** [comment]
Yep, we got some conflicts here.

**Comment by vaikas** [comment]
Flaked fetching release yaml from github:
```
 Install TUF
  error: unable to read URL "https://github.com/sigstore/scaffolding/releases/download/v0.4.8/release-tuf.yaml", server reported 503 Service Unavailable, status code=503
  Error: Process completed with exit code 1.
```

File: pkg/webhook/validator.go [hunk 1]
```diff
@@ -881,3 +823,104 @@ func getNamespace(ctx context.Context, namespace string) string {
...
+// contexts and the container could be nested at different levels in the
+// resource
+// index is the number in the containers array from the said context.
+//
+// Returns any encountered errors, or nil in two cases:
+// All the matched policies were validated, or
+// no matching policies were found, but the PolicyControllerConfig has been
+// configured to allow images not matching any policies.
+func (v *Validator) validateContainer(ctx context.Context, c corev1.Container, namespace, field string, index int, kind, apiVersion string, labels map[string]string, ociRemoteOpts ...ociremote.Option) *apis.FieldError {
+	ref, err := name.ParseReference(c.Image)
```
Comment (hectorj2f) [inline pkg/webhook/validator.go:839]: nit: one potential improvement could be avoiding to parse the reference twice, as it was done in `refOrFieldError(`.

**Review by hectorj2f — APPROVED** [review]
lgtm, i added a nit suggestion

File: pkg/webhook/validator.go (hunk 1 above)
Comment (vaikas) [reply pkg/webhook/validator.go:839]: Yeah, thought about that, but figured that adding one more field to the signature vs. just parsing it again was lesser of two evils :)

**Pull request merged**